# Before your start:
- Read the README.md file
- Comment as much as you can and use the resources in the README.md file
- Happy learning!

In [56]:
# Libraries
import math
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import scipy.stats as st
import statsmodels.api as sm
import statsmodels.formula.api as smf
pd.set_option('display.max_columns', None)
# pd.reset_option('display.max_columns')


# Challenge 1 - Import and Describe the Dataset

#### In this challenge we will use the `austin_weather` data. 

#### First, import it into a data frame called `austin`. 

In [57]:
# Your code here
austin=pd.read_csv("austin_weather.csv")
austin.head()

Date  TempHighF  TempAvgF  TempLowF DewPointHighF DewPointAvgF  \
0  2013-12-21         74        60        45            67           49   
1  2013-12-22         56        48        39            43           36   
2  2013-12-23         58        45        32            31           27   
3  2013-12-24         61        46        31            36           28   
4  2013-12-25         58        50        41            44           40   

  DewPointLowF HumidityHighPercent HumidityAvgPercent HumidityLowPercent  \
0           43                  93                 75                 57   
1           28                  93                 68                 43   
2           23                  76                 52                 27   
3           21                  89                 56                 22   
4           36                  86                 71                 56   

  SeaLevelPressureHighInches SeaLevelPressureAvgInches  \
0                      29.86                     29.68   
1                      30.41                     30.13   
2                      30.56                     30.49   
3                      30.56                     30.45   
4                      30.41                     30.33   

  SeaLevelPressureLowInches VisibilityHighMiles VisibilityAvgMiles  \
0                     29.59                  10                  7   
1                     29.87                  10                 10   
2                     30.41                  10                 10   
3                      30.3                  10                 10   
4                     30.27                  10                 10   

  VisibilityLowMiles WindHighMPH WindAvgMPH WindGustMPH  \
0                  2          20          4          31   
1                  5          16          6          25   
2                 10           8          3          12   
3                  7          12          4          20   
4                  7          10          2          16   

  PrecipitationSumInches               Events  
0                   0.46  Rain , Thunderstorm  
1                      0                       
2                      0                       
3                      0                       
4                      T

In [58]:
df=austin.copy()

#### Next, describe the dataset you have loaded: 
- Look at the variables and their types
- Examine the descriptive statistics of the numeric variables 
- Look at the first five rows of all variables to evaluate the categorical variables as well

In [59]:
# Your code here
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1319 entries, 0 to 1318
Data columns (total 21 columns):
 #   Column                      Non-Null Count  Dtype 
---  ------                      --------------  ----- 
 0   Date                        1319 non-null   object
 1   TempHighF                   1319 non-null   int64 
 2   TempAvgF                    1319 non-null   int64 
 3   TempLowF                    1319 non-null   int64 
 4   DewPointHighF               1319 non-null   object
 5   DewPointAvgF                1319 non-null   object
 6   DewPointLowF                1319 non-null   object
 7   HumidityHighPercent         1319 non-null   object
 8   HumidityAvgPercent          1319 non-null   object
 9   HumidityLowPercent          1319 non-null   object
 10  SeaLevelPressureHighInches  1319 non-null   object
 11  SeaLevelPressureAvgInches   1319 non-null   object
 12  SeaLevelPressureLowInches   1319 non-null   object
 13  VisibilityHighMiles         1319 non-null   obje

In [60]:
# Your code here
df.nunique()

Date                          1319
TempHighF                       74
TempAvgF                        64
TempLowF                        61
DewPointHighF                   64
DewPointAvgF                    66
DewPointLowF                    73
HumidityHighPercent             58
HumidityAvgPercent              69
HumidityLowPercent              82
SeaLevelPressureHighInches     105
SeaLevelPressureAvgInches      101
SeaLevelPressureLowInches      105
VisibilityHighMiles              5
VisibilityAvgMiles              10
VisibilityLowMiles              12
WindHighMPH                     22
WindAvgMPH                      13
WindGustMPH                     37
PrecipitationSumInches         114
Events                           9
dtype: int64

In [61]:
# Your code here
df.describe()

TempHighF     TempAvgF     TempLowF
count  1319.000000  1319.000000  1319.000000
mean     80.862775    70.642911    59.902957
std      14.766523    14.045904    14.190648
min      32.000000    29.000000    19.000000
25%      72.000000    62.000000    49.000000
50%      83.000000    73.000000    63.000000
75%      92.000000    83.000000    73.000000
max     107.000000    93.000000    81.000000

#### Given the information you have learned from examining the dataset, write down three insights about the data in a markdown cell below

#### Your Insights:

1. There are 21 variables in the dataset. 3 of them are numeric and the rest contain some text.

2. The average temperature in Austin ranged between around 70 degrees F and around 93 degrees F. The highest temperature observed during this period was 107 degrees F and the lowest was 19 degrees F.

3. When we look at the head function, we see that a lot of variables contain numeric data even though these columns are of object type. This means we might have to do some data cleansing.


#### Let's examine the DewPointAvgF variable by using the `unique()` function to list all unique values in this dataframe.

Describe what you find in a markdown cell below the code. What did you notice? What do you think made Pandas to treat this column as *object* instead of *int64*? 

In [62]:
# Your code here
df.DewPointAvgF.unique()

array(['49', '36', '27', '28', '40', '39', '41', '26', '42', '22', '48',
       '32', '8', '11', '45', '55', '61', '37', '47', '25', '23', '20',
       '33', '30', '29', '17', '14', '13', '54', '59', '15', '24', '34',
       '35', '57', '50', '53', '60', '46', '56', '51', '31', '38', '62',
       '43', '63', '64', '67', '66', '58', '70', '68', '65', '69', '71',
       '72', '-', '73', '74', '21', '44', '52', '12', '75', '76', '18'],
      dtype=object)

In [63]:
# Your observation here 
# There is a hyphen that is preventing pandas
# to convert the column to numeric type

The following is a list of columns misrepresented as `object`. Use this list to convert the columns to numeric using the `pandas.to_numeric` function in the next cell. If you encounter errors in converting strings to numeric values, you need to catch those errors and force the conversion by supplying `errors='coerce'` as an argument for `pandas.to_numeric`. Coercing will replace non-convertable elements with `NaN` which represents an undefined numeric value. This makes it possible for us to conveniently handle missing values in subsequent data processing.

*Hint: you may use a loop to change one column at a time but it is more efficient to use `apply`.*

In [64]:
wrong_type_columns = ['DewPointHighF', 'DewPointAvgF', 'DewPointLowF', 'HumidityHighPercent', 
                      'HumidityAvgPercent', 'HumidityLowPercent', 'SeaLevelPressureHighInches', 
                      'SeaLevelPressureAvgInches' ,'SeaLevelPressureLowInches', 'VisibilityHighMiles',
                      'VisibilityAvgMiles', 'VisibilityLowMiles', 'WindHighMPH', 'WindAvgMPH', 
                      'WindGustMPH', 'PrecipitationSumInches']

In [65]:
# Your code here
df1=df.copy()


#### Check if your code has worked by printing the data types again. You should see only two `object` columns (`Date` and `Events`) now. All other columns should be `int64` or `float64`.

In [66]:
# Your code here
df1[wrong_type_columns]=df1[wrong_type_columns].apply(pd.to_numeric, errors="coerce")

In [67]:
df1.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1319 entries, 0 to 1318
Data columns (total 21 columns):
 #   Column                      Non-Null Count  Dtype  
---  ------                      --------------  -----  
 0   Date                        1319 non-null   object 
 1   TempHighF                   1319 non-null   int64  
 2   TempAvgF                    1319 non-null   int64  
 3   TempLowF                    1319 non-null   int64  
 4   DewPointHighF               1312 non-null   float64
 5   DewPointAvgF                1312 non-null   float64
 6   DewPointLowF                1312 non-null   float64
 7   HumidityHighPercent         1317 non-null   float64
 8   HumidityAvgPercent          1317 non-null   float64
 9   HumidityLowPercent          1317 non-null   float64
 10  SeaLevelPressureHighInches  1316 non-null   float64
 11  SeaLevelPressureAvgInches   1316 non-null   float64
 12  SeaLevelPressureLowInches   1316 non-null   float64
 13  VisibilityHighMiles         1307 

# Challenge 2 - Handle the Missing Data

#### Now that we have fixed the type mismatch, let's address the missing data.

By coercing the columns to numeric, we have created `NaN` for each cell containing characters. We should choose a strategy to address these missing data.

The first step is to examine how many rows contain missing data.

We check how much missing data we have by applying the `.isnull()` function to our dataset. To find the rows with missing data in any of its cells, we apply `.any(axis=1)` to the function. `austin.isnull().any(axis=1)` will return a column containing true if the row contains at least one missing value and false otherwise. Therefore we must subset our dataframe with this column. This will give us all rows with at least one missing value. 

#### In the next cell, identify all rows containing at least one missing value. Assign the dataframes with missing values to a variable called `missing_values`.

In [68]:
df2=df1.copy()

In [69]:
# Your code here
df2["null"]=df2.isnull().any(axis=1)

In [80]:
df2.head()

Date  TempHighF  TempAvgF  TempLowF  DewPointHighF  DewPointAvgF  \
0  2013-12-21         74        60        45           67.0          49.0   
1  2013-12-22         56        48        39           43.0          36.0   
2  2013-12-23         58        45        32           31.0          27.0   
3  2013-12-24         61        46        31           36.0          28.0   
4  2013-12-25         58        50        41           44.0          40.0   

   DewPointLowF  HumidityHighPercent  HumidityAvgPercent  HumidityLowPercent  \
0          43.0                 93.0                75.0                57.0   
1          28.0                 93.0                68.0                43.0   
2          23.0                 76.0                52.0                27.0   
3          21.0                 89.0                56.0                22.0   
4          36.0                 86.0                71.0                56.0   

   SeaLevelPressureHighInches  SeaLevelPressureAvgInches  \
0                       29.86                      29.68   
1                       30.41                      30.13   
2                       30.56                      30.49   
3                       30.56                      30.45   
4                       30.41                      30.33   

   SeaLevelPressureLowInches  VisibilityHighMiles  VisibilityAvgMiles  \
0                      29.59                 10.0                 7.0   
1                      29.87                 10.0                10.0   
2                      30.41                 10.0                10.0   
3                      30.30                 10.0                10.0   
4                      30.27                 10.0                10.0   

   VisibilityLowMiles  WindHighMPH  WindAvgMPH  WindGustMPH  \
0                 2.0         20.0         4.0         31.0   
1                 5.0         16.0         6.0         25.0   
2                10.0          8.0         3.0         12.0   
3                 7.0         12.0         4.0         20.0   
4                 7.0         10.0         2.0         16.0   

   PrecipitationSumInches               Events   null  
0                    0.46  Rain , Thunderstorm  False  
1                    0.00                       False  
2                    0.00                       False  
3                    0.00                       False  
4                     NaN                        True

In [83]:
df2_null=df2[df2["null"]==True]
df2_null.head()

Date  TempHighF  TempAvgF  TempLowF  DewPointHighF  DewPointAvgF  \
4   2013-12-25         58        50        41           44.0          40.0   
6   2013-12-27         60        53        45           41.0          39.0   
7   2013-12-28         62        51        40           43.0          39.0   
42  2014-02-01         76        66        55           62.0          59.0   
51  2014-02-10         60        48        35           49.0          36.0   

    DewPointLowF  HumidityHighPercent  HumidityAvgPercent  HumidityLowPercent  \
4           36.0                 86.0                71.0                56.0   
6           37.0                 83.0                65.0                47.0   
7           33.0                 92.0                64.0                36.0   
42          41.0                 81.0                71.0                60.0   
51          30.0                 82.0                74.0                66.0   

    SeaLevelPressureHighInches  SeaLevelPressureAvgInches  \
4                        30.41                      30.33   
6                        30.46                      30.39   
7                        30.33                      30.17   
42                       29.91                      29.81   
51                       30.23                      30.15   

    SeaLevelPressureLowInches  VisibilityHighMiles  VisibilityAvgMiles  \
4                       30.27                 10.0                10.0   
6                       30.34                 10.0                 9.0   
7                       30.04                 10.0                10.0   
42                      29.75                 10.0                10.0   
51                      30.02                 10.0                 8.0   

    VisibilityLowMiles  WindHighMPH  WindAvgMPH  WindGustMPH  \
4                  7.0         10.0         2.0         16.0   
6                  7.0          7.0         1.0         11.0   
7                  7.0         10.0         2.0         14.0   
42                 9.0         14.0         6.0         26.0   
51                 4.0         15.0         9.0         23.0   

    PrecipitationSumInches Events  null  
4                      NaN         True  
6                      NaN         True  
7                      NaN         True  
42                     NaN   Rain  True  
51                     NaN   Rain  True

There are multiple strategies to handle missing data. Below lists the most common ones data scientists use:

* Removing all rows or all columns containing missing data. This is the simplest strategy. It may work in some cases but not others.

* Filling all missing values with a placeholder value. 
    * For categorical data, `0`, `-1`, and `9999` are some commonly used placeholder values. 
    * For continuous data, some may opt to fill all missing data with the mean. This strategy is not optimal since it can increase the fit of the model.

* Filling the values using some algorithm. 

#### In our case, we will use a hybrid approach which is to first remove the data that contain most missing values then fill in the rest of the missing values with the *linear interpolation* algorithm.

#### Next, count the number of rows of `austin` and `missing_values`.

In [84]:
# Your code here
len(df2_null)

136

#### Calculate the ratio of missing rows to total rows

In [85]:
# Your code here
len(df2_null)/len(df2)

0.10310841546626232

As you can see, there is a large proportion of missing data (over 10%). Perhaps we should evaluate which columns have the most missing data and remove those columns. For the remaining columns, we will perform a linear approximation of the missing data.

We can find the number of missing rows in each column using the `.isna()` function. We then chain the `.sum` function to the `.isna()` function and find the number of missing rows per column

In [92]:
# Your code here
df2.columns=df2.columns.str.lower().str.replace(" ","_")

In [93]:
df2.columns.to_list()

['date',
 'temphighf',
 'tempavgf',
 'templowf',
 'dewpointhighf',
 'dewpointavgf',
 'dewpointlowf',
 'humidityhighpercent',
 'humidityavgpercent',
 'humiditylowpercent',
 'sealevelpressurehighinches',
 'sealevelpressureavginches',
 'sealevelpressurelowinches',
 'visibilityhighmiles',
 'visibilityavgmiles',
 'visibilitylowmiles',
 'windhighmph',
 'windavgmph',
 'windgustmph',
 'precipitationsuminches',
 'events',
 'null']

In [170]:
df2.isna().sum()

date                            0
temphighf                       0
tempavgf                        0
templowf                        0
dewpointhighf                   7
dewpointavgf                    7
dewpointlowf                    7
humidityhighpercent             2
humidityavgpercent              2
humiditylowpercent              2
sealevelpressurehighinches      3
sealevelpressureavginches       3
sealevelpressurelowinches       3
visibilityhighmiles            12
visibilityavgmiles             12
visibilitylowmiles             12
windhighmph                     2
windavgmph                      2
windgustmph                     4
precipitationsuminches        124
events                          0
null                            0
dtype: int64

#### As you can see from the output, the majority of missing data is in one column called `PrecipitationSumInches`. What's the number of missing values in this column in ratio to its total number of rows?

In [120]:
# Your code here
df2.precipitationsuminches.isna().sum()/len(df2.precipitationsuminches)

0.09401061410159212

Almost 10% data missing! Therefore, we prefer to remove this column instead of filling its missing values. 

#### Remove this column from `austin` using the `.drop()` function. Use the `inplace=True` argument.

*Hints:*

* By supplying `inplace=True` to `drop()`, the original dataframe object will be changed in place and the function will return `None`. In contrast, if you don't supply `inplace=True`, which is equivalent to supplying `inplace=False` because `False` is the default value, the original dataframe object will be kept and the function returns a copy of the transformed dataframe object. In the latter case, you'll have to assign the returned object back to your variable.

* Also, since you are dropping a column instead of a row, you'll need to supply `axis=1` to `drop()`.

[Reference for `pandas.DataFrame.drop`](https://pandas.pydata.org/pandas-docs/stable/generated/pandas.DataFrame.drop.html)

In [125]:
# Your code here 
df3=df2.copy()
df3=df3.drop(columns="precipitationsuminches")
# Print `austin` to confirm the column is indeed removed

df3

date  temphighf  tempavgf  templowf  dewpointhighf  dewpointavgf  \
0     2013-12-21         74        60        45           67.0          49.0   
1     2013-12-22         56        48        39           43.0          36.0   
2     2013-12-23         58        45        32           31.0          27.0   
3     2013-12-24         61        46        31           36.0          28.0   
4     2013-12-25         58        50        41           44.0          40.0   
...          ...        ...       ...       ...            ...           ...   
1314  2017-07-27        103        89        75           71.0          67.0   
1315  2017-07-28        105        91        76           71.0          64.0   
1316  2017-07-29        107        92        77           72.0          64.0   
1317  2017-07-30        106        93        79           70.0          68.0   
1318  2017-07-31         99        88        77           66.0          61.0   

      dewpointlowf  humidityhighpercent  humidityavgpercent  \
0             43.0                 93.0                75.0   
1             28.0                 93.0                68.0   
2             23.0                 76.0                52.0   
3             21.0                 89.0                56.0   
4             36.0                 86.0                71.0   
...            ...                  ...                 ...   
1314          61.0                 82.0                54.0   
1315          55.0                 87.0                54.0   
1316          55.0                 82.0                51.0   
1317          63.0                 69.0                48.0   
1318          54.0                 64.0                43.0   

      humiditylowpercent  sealevelpressurehighinches  \
0                   57.0                       29.86   
1                   43.0                       30.41   
2                   27.0                       30.56   
3                   22.0                       30.56   
4                   56.0                       30.41   
...                  ...                         ...   
1314                25.0                       30.04   
1315                20.0                       29.97   
1316                19.0                       29.91   
1317                27.0                       29.96   
1318                22.0                       30.04   

      sealevelpressureavginches  sealevelpressurelowinches  \
0                         29.68                      29.59   
1                         30.13                      29.87   
2                         30.49                      30.41   
3                         30.45                      30.30   
4                         30.33                      30.27   
...                         ...                        ...   
1314                      29.97                      29.88   
1315                      29.90                      29.81   
1316                      29.86                      29.79   
1317                      29.91                      29.87   
1318                      29.97                      29.91   

      visibilityhighmiles  visibilityavgmiles  visibilitylowmiles  \
0                    10.0                 7.0                 2.0   
1                    10.0                10.0                 5.0   
2                    10.0                10.0                10.0   
3                    10.0                10.0                 7.0   
4                    10.0                10.0                 7.0   
...                   ...                 ...                 ...   
1314                 10.0                10.0                10.0   
1315                 10.0                10.0                10.0   
1316                 10.0                10.0                10.0   
1317                 10.0                10.0                10.0   
1318                 10.0                10.0                10.0   

      windhighmph  windavgmph  windgustmph     

In [133]:
df3.events.eq(" ").sum() ## OJO CUIDAO, 
                        ## Tendremos que limpiar
                        ## cuando lo mande el lab

903

#### Next we will perform linear interpolation of the missing data.

This means that we will use a linear algorithm to estimate the missing data. Linear interpolation assumes that there is a straight line between the points and the missing point will fall on that line. This is a good enough approximation for weather related data. Weather related data is typically a time series. Therefore, we do not want to drop rows from our data if possible. It is prefereable to estimate the missing values rather than remove the rows. However, if you have data from a single point in time, perhaps a better solution would be to remove the rows. 

If you would like to read more about linear interpolation, you can do so [here](https://en.wikipedia.org/wiki/Linear_interpolation).

In the following cell, use the `.interpolate()` function on the entire dataframe. This time pass the `inplace=False` argument to the function and assign the interpolated dataframe to a new variable called `austin_fixed` so that we can compare with `austin`.

In [135]:
# Your code here
df4=df3.copy()
df4_fixed=df4.interpolate()

/tmp/ipykernel_32932/4278860520.py:3: FutureWarning: DataFrame.interpolate with object dtype is deprecated and will raise in a future version. Call obj.infer_objects(copy=False) before interpolating instead.
  df4_fixed=df4.interpolate()


#### Check to make sure `austin_fixed` contains no missing data. Also check `austin` - it still contains missing data.

In [139]:
# Your code here
df4_fixed.isnull().any(axis=1).sum()

0

In [140]:
df4.isnull().any(axis=1).sum()

14

# Challenge 3 - Processing the `Events` Column

#### Our dataframe contains one true text column - the Events column. We should evaluate this column to determine how to process it.

Use the `value_counts()` function to evaluate the contents of this column

In [230]:
# Your code here:
df5=df4_fixed.copy()

Reading the values of `Events` and reflecting what those values mean in the context of data, you realize this column indicates what weather events had happened in a particular day.

#### What is the largest number of events happened in a single day? Enter your answer in the next cell.

In [231]:
# Your answer:
df5.events.value_counts()

events
                             903
Rain                         192
Rain , Thunderstorm          137
Fog , Rain , Thunderstorm     33
Fog                           21
Thunderstorm                  17
Fog , Rain                    14
Rain , Snow                    1
Fog , Thunderstorm             1
Name: count, dtype: int64

#### We want to transform the string-type `Events` values to the numbers. This will allow us to apply machine learning algorithms easily.

How? We will create a new column for each type of events (i.e. *Rain*, *Snow*, *Fog*, *Thunderstorm*. In each column, we use `1` to indicate if the corresponding event happened in that day and use `0` otherwise.

Below we provide you a list of all event types. Loop the list and create a dummy column with `0` values for each event in `austin_fixed`. To create a new dummy column with `0` values, simply use `austin_fixed[event] = 0`.

In [232]:
# df5["events"]=df5["events"].replace(" ",0)
# df5["events"] = df5["events"].str.split(", ")


In [233]:
df6=df5.copy()

In [234]:
event_list = ['snow', 'fog', 'rain', 'thunderstorm']

# Your code here
for event in event_list:
    df6[event]=0

#### Next, populate the actual values in the dummy columns of  `austin_fixed`.

You will check the *Events* column. If its string value contains `Rain`, then the *Rain* column should be `1`. The same for `Snow`, `Fog`, and `Thunderstorm`.

*Hints:*

* Use [`pandas.Series.str.contains()`](https://pandas.pydata.org/pandas-docs/stable/generated/pandas.Series.str.contains.html) to create the value series of each new column.

* What if the values you populated are booleans instead of numbers? You can cast the boolean values to numbers by using `.astype(int)`. For instance, `pd.Series([True, True, False]).astype(int)` will return a new series with values of `[1, 1, 0]`.

#### Print out `austin_fixed` to check if the event columns are populated with the intended values

In [247]:
df6["rain"] = df6["events"].str.contains("Rain", na=False).astype(int)
df6["snow"] = df6["events"].str.contains("Snow", na=False).astype(int)
df6["fog"] = df6["events"].str.contains("Fog", na=False).astype(int)
df6["thunderstorm"] = df6["events"].str.contains("Thunderstorm", na=False).astype(int)
df6.rain=df6.rain.fillna(0)
df6.snow=df6.snow.fillna(0)
df6.fog=df6.fog.fillna(0)
df6.thunderstorm=df6.thunderstorm.fillna(0)

In [248]:
df6.head()

date  temphighf  tempavgf  templowf  dewpointhighf  dewpointavgf  \
0  2013-12-21         74        60        45           67.0          49.0   
1  2013-12-22         56        48        39           43.0          36.0   
2  2013-12-23         58        45        32           31.0          27.0   
3  2013-12-24         61        46        31           36.0          28.0   
4  2013-12-25         58        50        41           44.0          40.0   

   dewpointlowf  humidityhighpercent  humidityavgpercent  humiditylowpercent  \
0          43.0                 93.0                75.0                57.0   
1          28.0                 93.0                68.0                43.0   
2          23.0                 76.0                52.0                27.0   
3          21.0                 89.0                56.0                22.0   
4          36.0                 86.0                71.0                56.0   

   sealevelpressurehighinches  sealevelpressureavginches  \
0                       29.86                      29.68   
1                       30.41                      30.13   
2                       30.56                      30.49   
3                       30.56                      30.45   
4                       30.41                      30.33   

   sealevelpressurelowinches  visibilityhighmiles  visibilityavgmiles  \
0                      29.59                 10.0                 7.0   
1                      29.87                 10.0                10.0   
2                      30.41                 10.0                10.0   
3                      30.30                 10.0                10.0   
4                      30.27                 10.0                10.0   

   visibilitylowmiles  windhighmph  windavgmph  windgustmph  \
0                 2.0         20.0         4.0         31.0   
1                 5.0         16.0         6.0         25.0   
2                10.0          8.0         3.0         12.0   
3                 7.0         12.0         4.0         20.0   
4                 7.0         10.0         2.0         16.0   

                events   null  snow  fog  rain  thunderstorm  
0  Rain , Thunderstorm  False     0    0     1             1  
1                       False     0    0     0             0  
2                       False     0    0     0             0  
3                       False     0    0     0             0  
4                        True     0    0     0             0

In [249]:
## Solución de Fer. No me va 
# a funcionar porque estandaricé columnas y los eventos
# (valores) est√°n capitalizados pero el nombre de 
# columna, no. 
# for event in event_list:
#     df6[event] = df6["events"].str.contains(event,na=False).astype(int)

#### If your code worked correctly, now we can drop the `Events` column as we don't need it any more.

In [250]:
df6.head()

date  temphighf  tempavgf  templowf  dewpointhighf  dewpointavgf  \
0  2013-12-21         74        60        45           67.0          49.0   
1  2013-12-22         56        48        39           43.0          36.0   
2  2013-12-23         58        45        32           31.0          27.0   
3  2013-12-24         61        46        31           36.0          28.0   
4  2013-12-25         58        50        41           44.0          40.0   

   dewpointlowf  humidityhighpercent  humidityavgpercent  humiditylowpercent  \
0          43.0                 93.0                75.0                57.0   
1          28.0                 93.0                68.0                43.0   
2          23.0                 76.0                52.0                27.0   
3          21.0                 89.0                56.0                22.0   
4          36.0                 86.0                71.0                56.0   

   sealevelpressurehighinches  sealevelpressureavginches  \
0                       29.86                      29.68   
1                       30.41                      30.13   
2                       30.56                      30.49   
3                       30.56                      30.45   
4                       30.41                      30.33   

   sealevelpressurelowinches  visibilityhighmiles  visibilityavgmiles  \
0                      29.59                 10.0                 7.0   
1                      29.87                 10.0                10.0   
2                      30.41                 10.0                10.0   
3                      30.30                 10.0                10.0   
4                      30.27                 10.0                10.0   

   visibilitylowmiles  windhighmph  windavgmph  windgustmph  \
0                 2.0         20.0         4.0         31.0   
1                 5.0         16.0         6.0         25.0   
2                10.0          8.0         3.0         12.0   
3                 7.0         12.0         4.0         20.0   
4                 7.0         10.0         2.0         16.0   

                events   null  snow  fog  rain  thunderstorm  
0  Rain , Thunderstorm  False     0    0     1             1  
1                       False     0    0     0             0  
2                       False     0    0     0             0  
3                       False     0    0     0             0  
4                        True     0    0     0             0

In [ ]:
# Your code here
df6=df6.drop(columns="events")


KeyError: "['events'] not found in axis"

In [253]:

df6.head()

date  temphighf  tempavgf  templowf  dewpointhighf  dewpointavgf  \
0  2013-12-21         74        60        45           67.0          49.0   
1  2013-12-22         56        48        39           43.0          36.0   
2  2013-12-23         58        45        32           31.0          27.0   
3  2013-12-24         61        46        31           36.0          28.0   
4  2013-12-25         58        50        41           44.0          40.0   

   dewpointlowf  humidityhighpercent  humidityavgpercent  humiditylowpercent  \
0          43.0                 93.0                75.0                57.0   
1          28.0                 93.0                68.0                43.0   
2          23.0                 76.0                52.0                27.0   
3          21.0                 89.0                56.0                22.0   
4          36.0                 86.0                71.0                56.0   

   sealevelpressurehighinches  sealevelpressureavginches  \
0                       29.86                      29.68   
1                       30.41                      30.13   
2                       30.56                      30.49   
3                       30.56                      30.45   
4                       30.41                      30.33   

   sealevelpressurelowinches  visibilityhighmiles  visibilityavgmiles  \
0                      29.59                 10.0                 7.0   
1                      29.87                 10.0                10.0   
2                      30.41                 10.0                10.0   
3                      30.30                 10.0                10.0   
4                      30.27                 10.0                10.0   

   visibilitylowmiles  windhighmph  windavgmph  windgustmph   null  snow  fog  \
0                 2.0         20.0         4.0         31.0  False     0    0   
1                 5.0         16.0         6.0         25.0  False     0    0   
2                10.0          8.0         3.0         12.0  False     0    0   
3                 7.0         12.0         4.0         20.0  False     0    0   
4                 7.0         10.0         2.0         16.0   True     0    0   

   rain  thunderstorm  
0     1             1  
1     0             0  
2     0             0  
3     0             0  
4     0             0

# Challenge 4 - Processing The `Date` Column

The `Date` column is another non-numeric field in our dataset. A value in that field looks like `'2014-01-06'` which consists of the year, month, and day connected with hyphens. One way to convert the date string to numerical is using a similar approach as we used for `Events`, namely splitting the column into numerical `Year`, `Month`, and `Day` columns. In this challenge we'll show you another way which is to use the Python `datetime` library's `toordinal()` function. Depending on what actual machine learning analysis you will conduct, each approach has its pros and cons. Our goal today is to practice data preparation so we'll skip the discussion here.

Here you can find the [reference](https://docs.python.org/3/library/datetime.html) and [example](https://stackoverflow.com/questions/39846918/convert-date-to-ordinal-python) for `toordinal`. The basic process is to first convert the string to a `datetime` object using `datetime.datetime.strptime`, then convert the `datetime` object to numerical using `toordinal`.

#### In the cell below, convert the `Date` column values from string to numeric values using `toordinal()`.

In [ ]:
# Your code here
df7=df6.copy()
df7.date=pd.to_datetime(df7.date)

In [274]:
## TO DATETIME FUNCIONA POR COLUMNA PERO TOORDINAL
## FUNCIONA LINEA A LINEA, POR ESO LE TENEMOS QUE
## PASAR UN APPLY LAMBDA
from datetime import datetime
df7.date=df7['date'].apply(lambda x: x.toordinal())


AttributeError: 'int' object has no attribute 'toordinal'

#### Print `austin_fixed` to check your `Date` column.

In [275]:

df7.head()

date  temphighf  tempavgf  templowf  dewpointhighf  dewpointavgf  \
0  735223         74        60        45           67.0          49.0   
1  735224         56        48        39           43.0          36.0   
2  735225         58        45        32           31.0          27.0   
3  735226         61        46        31           36.0          28.0   
4  735227         58        50        41           44.0          40.0   

   dewpointlowf  humidityhighpercent  humidityavgpercent  humiditylowpercent  \
0          43.0                 93.0                75.0                57.0   
1          28.0                 93.0                68.0                43.0   
2          23.0                 76.0                52.0                27.0   
3          21.0                 89.0                56.0                22.0   
4          36.0                 86.0                71.0                56.0   

   sealevelpressurehighinches  sealevelpressureavginches  \
0                       29.86                      29.68   
1                       30.41                      30.13   
2                       30.56                      30.49   
3                       30.56                      30.45   
4                       30.41                      30.33   

   sealevelpressurelowinches  visibilityhighmiles  visibilityavgmiles  \
0                      29.59                 10.0                 7.0   
1                      29.87                 10.0                10.0   
2                      30.41                 10.0                10.0   
3                      30.30                 10.0                10.0   
4                      30.27                 10.0                10.0   

   visibilitylowmiles  windhighmph  windavgmph  windgustmph   null  snow  fog  \
0                 2.0         20.0         4.0         31.0  False     0    0   
1                 5.0         16.0         6.0         25.0  False     0    0   
2                10.0          8.0         3.0         12.0  False     0    0   
3                 7.0         12.0         4.0         20.0  False     0    0   
4                 7.0         10.0         2.0         16.0   True     0    0   

   rain  thunderstorm  
0     1             1  
1     0             0  
2     0             0  
3     0             0  
4     0             0

# Challenge 5 - Sampling and Holdout Sets

#### Now that we have processed the data for machine learning, we will separate the data to test and training sets.

We first train the model using only the training set. We check our metrics on the training set. We then apply the model to the test set and check our metrics on the test set as well. If the metrics are significantly more optimal on the training set, then we know we have overfit our model. We will need to revise our model to ensure it will be more applicable to data outside the test set.

#### In the next cells we will separate the data into a training set and a test set using the `train_test_split()` function in scikit-learn.

When using `scikit-learn` for machine learning, we first separate the data to predictor and response variables. This is the standard way of passing datasets into a model in `scikit-learn`. The `scikit-learn` will then find out whether the predictors and responses fit the model.

In the next cell, assign the `TempAvgF` column to `y` and the remaining columns to `X`. Your `X` should be a subset of `austin_fixed` containing the following columns: 

```['Date',
 'TempHighF',
 'TempLowF',
 'DewPointHighF',
 'DewPointAvgF',
 'DewPointLowF',
 'HumidityHighPercent',
 'HumidityAvgPercent',
 'HumidityLowPercent',
 'SeaLevelPressureHighInches',
 'SeaLevelPressureAvgInches',
 'SeaLevelPressureLowInches',
 'VisibilityHighMiles',
 'VisibilityAvgMiles',
 'VisibilityLowMiles',
 'WindHighMPH',
 'WindAvgMPH',
 'WindGustMPH',
 'Snow',
 'Fog',
 'Rain',
 'Thunderstorm']```
 
 Your `y` should be a subset of `austin_fixed` containing one column `TempAvgF`.

In [293]:
# Your code here:
y=df7.tempavgf
X=df7.drop(columns="tempavgf")

In the next cell, import `train_test_split` from `sklearn.model_selection`

In [277]:
# 🤖 Machine Learning
from sklearn.model_selection import train_test_split # ‚ùó New

Now that we have split the data to predictor and response variables and imported the `train_test_split()` function, split `X` and `y` into `X_train`, `X_test`, `y_train`, and `y_test`. 80% of the data should be in the training set and 20% in the test set. `train_test_split()` reference can be accessed [here](https://scikit-learn.org/stable/modules/generated/sklearn.model_selection.train_test_split.html).


Enter your code in the cell below:

In [278]:
#Your code here:
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=42)

#### Congratulations! Now you have finished the preparation of the dataset!

# Bonus Challenge 1

#### While the above is the common practice to prepare most datasets, when it comes to time series data, we sometimes do not want to randomly select rows from our dataset.

This is because many time series algorithms rely on observations having equal time distances between them. In such cases, we typically select the majority of rows as the test data and the last few rows as the training data. We don't use `train_test_split()` to select the train/test data because it returns random selections.

In the following cell, compute the number of rows that account for 80% of our data and round it to the next integer. Assign this number to `ts_rows`.

In [292]:
# Your code here:
ts_rows=round(len(df7)/100*80)
ts_rows

1055

Assign the first `ts_rows` rows of `X` to `X_ts_train` and the remaining rows to `X_ts_test`.

Assign the first `ts_rows` rows of `y` to `y_ts_train` and the remaining rows to `y_ts_test`.

Hay un parámetro en train_test_split que es suffle=False, que si lo ponemos así, impide la randomización de la toma del test y del train. Pero intentaremos hacerlo según aquí se especifica.


In [294]:
# Your code here:
X_ts_train, X_ts_test, y_ts_train, y_ts_test = train_test_split(X, y, test_size=0.2, shuffle=False)

In [295]:
X_ts_train

date  temphighf  templowf  dewpointhighf  dewpointavgf  dewpointlowf  \
0     735223         74        45           67.0          49.0          43.0   
1     735224         56        39           43.0          36.0          28.0   
2     735225         58        32           31.0          27.0          23.0   
3     735226         61        31           36.0          28.0          21.0   
4     735227         58        41           44.0          40.0          36.0   
...      ...        ...       ...            ...           ...           ...   
1050  736273         72        63           68.0          66.0          59.0   
1051  736274         76        59           67.0          60.0          57.0   
1052  736275         71        61           69.0          65.0          61.0   
1053  736276         69        62           66.0          64.0          62.0   
1054  736277         65        60           62.0          61.0          58.0   

      humidityhighpercent  humidityavgpercent  humiditylowpercent  \
0                    93.0                75.0                57.0   
1                    93.0                68.0                43.0   
2                    76.0                52.0                27.0   
3                    89.0                56.0                22.0   
4                    86.0                71.0                56.0   
...                   ...                 ...                 ...   
1050                100.0                91.0                81.0   
1051                 93.0                85.0                76.0   
1052                100.0                94.0                87.0   
1053                100.0                92.0                84.0   
1054                100.0                94.0                87.0   

      sealevelpressurehighinches  sealevelpressureavginches  \
0                          29.86                      29.68   
1                          30.41                      30.13   
2                          30.56                      30.49   
3                          30.56                      30.45   
4                          30.41                      30.33   
...                          ...                        ...   
1050                       30.27                      30.22   
1051                       30.19                      30.13   
1052                       30.14                      30.09   
1053                       30.21                      30.16   
1054                       30.29                      30.23   

      sealevelpressurelowinches  visibilityhighmiles  visibilityavgmiles  \
0                         29.59                 10.0                 7.0   
1                         29.87                 10.0                10.0   
2                         30.41                 10.0                10.0   
3                         30.30                 10.0                10.0   
4                         30.27                 10.0                10.0   
...                         ...                  ...                 ...   
1050                      30.14                 10.0                 6.0   
1051                      30.07                 10.0                10.0   
1052                      30.04                 10.0                 5.0   
1053                      30.09                 10.0                 8.0   
1054                      30.18                 10.0                 6.0   

      visibilitylowmiles  windhighmph  windavgmph  windgustmph   null  snow  \
0                    2.0         20.0         4.0         31.0  False     0   
1                    5.0         16.0         6.0         25.0  False     0   
2                   10.0          8.0         3.0         12.0  False     0   
3                    7.0         12.0         4.0         20.0  False     0   
4                    7.0         10.0         2.0         16.0   True     0   
...                  ...          ...         ...          ...    ...   ..

In [296]:
## Método diana/ fer

X_ts_train_2=X.iloc[:ts_rows]
X_ts_test_2=X.iloc[ts_rows:]
y_ts_train_2=y.iloc[:ts_rows]
y_ts_test_2=y.iloc[ts_rows]

In [297]:
X_ts_train==X_ts_train_2

date  temphighf  templowf  dewpointhighf  dewpointavgf  dewpointlowf  \
0     True       True      True           True          True          True   
1     True       True      True           True          True          True   
2     True       True      True           True          True          True   
3     True       True      True           True          True          True   
4     True       True      True           True          True          True   
...    ...        ...       ...            ...           ...           ...   
1050  True       True      True           True          True          True   
1051  True       True      True           True          True          True   
1052  True       True      True           True          True          True   
1053  True       True      True           True          True          True   
1054  True       True      True           True          True          True   

      humidityhighpercent  humidityavgpercent  humiditylowpercent  \
0                    True                True                True   
1                    True                True                True   
2                    True                True                True   
3                    True                True                True   
4                    True                True                True   
...                   ...                 ...                 ...   
1050                 True                True                True   
1051                 True                True                True   
1052                 True                True                True   
1053                 True                True                True   
1054                 True                True                True   

      sealevelpressurehighinches  sealevelpressureavginches  \
0                           True                       True   
1                           True                       True   
2                           True                       True   
3                           True                       True   
4                           True                       True   
...                          ...                        ...   
1050                        True                       True   
1051                        True                       True   
1052                        True                       True   
1053                        True                       True   
1054                        True                       True   

      sealevelpressurelowinches  visibilityhighmiles  visibilityavgmiles  \
0                          True                 True                True   
1                          True                 True                True   
2                          True                 True                True   
3                          True                 True                True   
4                          True                 True                True   
...                         ...                  ...                 ...   
1050                       True                 True                True   
1051                       True                 True                True   
1052                       True                 True                True   
1053                       True                 True                True   
1054                       True                 True                True   

      visibilitylowmiles  windhighmph  windavgmph  windgustmph  null  snow  \
0                   True         True        True         True  True  True   
1                   True         True        True         True  True  True   
2                   True         True        True         True  True  True   
3                   True         True        True         True  True  True   
4                   True         True        True         True  True  True   
...                  ...          ...         ...          ...   ...   ...   
1050                True

Hemos comprobado que ambos métodos funcionan bien 